File: 121_best_time_to_buy_and_sell_stock/one_pass.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int ans = 0;
        int min_val = 32767;
        int tmp_diff;
        for(int i = 0; i < prices.size(); ++i){
            if(prices[i] < min_val) min_val = prices[i]; 
            tmp_diff = prices[i] - min_val;
            if(tmp_diff > ans) ans = tmp_diff;
        }
        return ans;
    }
};

File: 121_best_time_to_buy_and_sell_stock/one_pass.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """ 
        """
        Complexity: O(n)
        Space: O(1)
        """
        ans = 0
        min_val = float('inf')
        for i in range(len(prices)):
            if prices[i] < min_val:
                min_val = prices[i]
            if prices[i] - min_val > ans:
                ans = prices[i] - min_val
        return ans
